### PR TITLE
Expose cutadapt `--cut` option

### PR DESCRIPTION
**Pull request by vaamb**
Allow to use cutadapt `--cut` option

This is useful in combination with anchored adapters (https://cutadapt.readthedocs.io/en/stable/guide.html#anchored-5adapters) when a poly-N is added before the tag/index/primer to increase Illumina sequences discrimination during its clustering phase

A follow-up pull request that adds an "anchored" mode will follow if this request is accepted

### CODE REVIEW

**Comment by colinvwood** [comment]
Hello @vaamb,

Thank you for your work on this. My understanding is that `--cut` looks only in the forward read in the case of paired end reads. In other words, we're not providing support for libraries that have their barcodes in the reverse reads (I know this is less common), nor for libraries that are dual-indexed. I'm wondering if we should make the help text more explicit about this.

I'm also wondering if `anchor_adapter` should be `anchor_barcode` instead since we're not really looking for and trimming adapters--that's a separate QC step. Let me know if there's something I'm missing.

**Comment by lizgehret** [comment]
Hey @vaamb - we'll need to bump this to next release if we don't hear back from you by EOD tomorrow. Thanks!

**Comment by vaamb** [comment]
Hello @colinvwood,

Sorry for the delay, I finally got some time to go back on this. 

Indeed `--cut` (= `-u`) is looking on the forward read only.
For the dual-indexed libraries, after looking through the `cutadapt` documentation I found the `-U` parameter which does the same as `-u` but on the reverse read. I changed the code so if paired-reads with forward and reverse barcodes are provided both `-u` and `-U` will be used.
For the case where the barcodes are in the reverse reads, do you mean only in the reverse reads or both in forward and reverse reads (mixed orientation)? In the former case, shouldn't the user switch its forward and reverse reads during import ? In the later case, there are three imperfect solutions:
- We apply the cut both on forward and reverse reads (as it is done for dual-index) during the first dumux step but the user will lose some nucleotides on both its forward and reverse reads and I should add a warning for this.
- We apply the cut on the forward reads during the first demux step, then after switching the forward and reverse reads, we redo the same on the "new" forward reads (same issue as for the first solution).
- We raise an exception and unable the use of `cut` in case of paired-reads with only a single barcode provided.

Concerning the nomenclature issue, I used "adapter" as it is the term used throughout the cutadapt documentation but I agree that "barcode" is more ... adapted.

**Comment by vaamb** [comment]
Rebased the branch on master to try to fix the CI failure during the `Create conda build config` phase

**Comment by colinvwood** [comment]
Hello @vaamb,

> We apply the cut both on forward and reverse reads (as it is done for dual-index) during the first dumux step but the user will lose some nucleotides on both its forward and reverse reads and I should add a warning for this.

Why would a warning be necessary here? Because it could possibly be undesired behavior from the user's point of view? We could expose two different cut values for each of the read directions if that would solve this problem.

> We apply the cut on the forward reads during the first demux step, then after switching the forward and reverse reads, we redo the same on the "new" forward reads (same issue as for the first solution).

This only comes into play for "mixed orientation" reads, right? I think the logic you have now makes good sense. 

> We raise an exception and unable the use of cut in case of paired-reads with only a single barcode provided.

My response to the first point would also solve this problem. 

So maybe the step forward is to expose two different cut parameters, both optional and independent of one another, remembering to apply them correctly (ie repeatedly) in the case of "mixed orientation" reads. What do you think about this?

I would also want to see some tests for a few other things (default of cut=0, cut applied to both read directions, and a negative cut value) before merging. I'm happy to help with these if you don't have the time.

**Comment by vaamb** [comment]
Hi @colinvwood,

Indeed exposing two `cut` parameters for `demux_paired` would be a better and more explicit solution!
I'll just raise an error if `forward_cut` and `reverse_cut` are not the same when using mixed orientation. As mixed orientation is simply doing a double pass demux, sequences will be cut during the first pass, then forward and reverse sequences swap and the second demux pass will be done without cut (as it was done during the first pass). As forward and reverse sequences (and barcodes) are interchangeable, their cut options should also be.

Concerning the tests, the `_build_demux_command` now uses `cut=0` by default (and adds a `'-u', str(cut)` to the command) and all the tests pass. 
For the other tests, can I simply adapt some current test as I did in `TestDemuxSingle.test_cut()`? Knowing this adaptation 'plays' with the fact that barcodes are not anchored by default in `cutadapt`? Therefore when I cut the first 'C' of the sequence 'CCCCacgtacgt' (giving 'CCCacgtacgt') it is still demuxed with the barcode 'CCCC'. Similarly, all the tests still pass when an arbitrarily long sequence is added at the 5' end of the barcode. I did some trials, and the tests still pass if the barcode is longer than the sequence, as long as the last 3 nucleotides of the barcode match the first 3 nucleotides of the sequence (this 3 is set with `cutadapt`'s `--overlap` option).

**Comment by colinvwood** [comment]
Hi @vaamb,

> Indeed exposing two cut parameters for demux_paired would be a better and more explicit solution!
I'll just raise an error if forward_cut and reverse_cut are not the same when using mixed orientation. As mixed orientation is simply doing a double pass demux, sequences will be cut during the first pass, then forward and reverse sequences swap and the second demux pass will be done without cut (as it was done during the first pass). As forward and reverse sequences (and barcodes) are interchangeable, their cut options should also be.

This sounds good to me.

> Concerning the tests, the _build_demux_command now uses cut=0 by default (and adds a '-u', str(cut) to the command) and all the tests pass.

Yep, duh. Sorry about that.

> For the other tests, can I simply adapt some current test as I did in TestDemuxSingle.test_cut()?

I don't see why not.

> Similarly, all the tests still pass when an arbitrarily long sequence is added at the 5' end of the barcode. I did some trials, and the tests still pass if the barcode is longer than the sequence, as long as the last 3 nucleotides of the barcode match the first 3 nucleotides of the sequence (this 3 is set with cutadapt's --overlap option).

This is intended behavior according to the cutadapt docs, so I think that's all alright. We don't expose the `--overlap` parameter, but there's an "ADAPTER;min_overlap=n" syntax that you could use in the tests if you want to play with that part.

Maybe some we should have some tests that cut enough to cause the demuxing of a sequence to fail? 

BTW: I'm not ignoring #58, but figured we would get this one merged and then start looking at that one.

**Comment by vaamb** [comment]
@colinvwood 

I split the `cut` option into `forward_cut` and `reverse_cut` and added some tests. If you see any other tests to add I can write them.

File: q2_cutadapt/plugin_setup.py [hunk 1]
```diff
@@ -334,6 +342,18 @@
                             'per-sample barcodes for the forward reads.',
         'reverse_barcodes': 'The sample metadata column listing the '
                             'per-sample barcodes for the reverse reads.',
+        'forward_cut': 'Remove the specified number of bases from the forward '
+                       'sequences. Bases are removed before demultiplexing. '
+                       'If a positive value is provided, bases are removed '
+                       'from the beginning of the sequences. If a negative '
+                       'value is provided, bases are removed from the end of '
+                       'the sequences',
+        'reverse_cut': 'Remove the specified number of bases from the forward '
```
Comment (colinvwood) [inline q2_cutadapt/plugin_setup.py]:
```suggestion
        'reverse_cut': 'Remove the specified number of bases from the reverse '
```

**Comment by colinvwood** [comment]
Hi @vaamb,

This looks good to me, thanks for all your work. I think the last thing we need are some tests for the interaction between `cut` and `mixed_orientation`. Maybe one test to make sure double cutting isn't happening and ~~one that raises the error condition~~.

One small issue that has been introduced is when paired reads with mixed orientation are given a forward cut value and not a reverse. This will raise an error but is unexpected from a user's point of view because the command line interface will indicate that both are optional (when really they are dependent on one another in the case of mixed orientation). We don't have any way to communicate this programmatically with click as far as I know, so I think we should add another explanatory line in the help text for both `forward_cut` and `reverse_cut` along the lines of:

> If `--p-mixed-orientation` is set, then both `--p-forward-cut` and `--p-reverse-cut` must both be set to the same value.

Or whatever you think is clearest. Let me know if you have a different idea about how to deal with this. 

EDIT: just saw that you already have a test for the error condition, so disregard.

**Comment by colinvwood** [comment]
Hi @vaamb,

One more thing: would you mind throwing in the additional asserts for forward_cut and reverse_cut in `TestDemuxUtilsPairedEnd::test_build_demux_command`?

**Comment by vaamb** [comment]
Hi @colinvwood 

Totally forgot to add this test in `TestDemuxUtilsPairedEnd`. It is now done.

File: q2_cutadapt/plugin_setup.py [hunk 2]
```diff
@@ -334,6 +342,22 @@
                             'per-sample barcodes for the forward reads.',
         'reverse_barcodes': 'The sample metadata column listing the '
                             'per-sample barcodes for the reverse reads.',
+        'forward_cut': 'Remove the specified number of bases from the forward '
+                       'sequences. Bases are removed before demultiplexing. '
+                       'If a positive value is provided, bases are removed '
+                       'from the beginning of the sequences. If a negative '
+                       'value is provided, bases are removed from the end of '
+                       'the sequences. If --p-mixed-orientation is set, then '
+                       'both --p-forward-cut and --p-reverse-cut must both be '
```
Comment (colinvwood) [inline q2_cutadapt/plugin_setup.py]:
```suggestion
                       'both --p-forward-cut and --p-reverse-cut must be '
```

Comment (colinvwood) [reply q2_cutadapt/plugin_setup.py]: (my redundant typo)

File: q2_cutadapt/plugin_setup.py [hunk 3]
```diff
@@ -334,6 +342,22 @@
...
+                       'set to the same value.',
+        'reverse_cut': 'Remove the specified number of bases from the reverse '
+                       'sequences. Bases are removed before demultiplexing. '
+                       'If a positive value is provided, bases are removed '
+                       'from the beginning of the sequences. If a negative '
+                       'value is provided, bases are removed from the end of '
+                       'the sequences. If --p-mixed-orientation is set, then '
+                       'both --p-forward-cut and --p-reverse-cut must both be '
```
Comment (colinvwood) [inline q2_cutadapt/plugin_setup.py]:
```suggestion
                       'both --p-forward-cut and --p-reverse-cut must be '
```

File: q2_cutadapt/tests/test_demux.py [hunk 4]
```diff
@@ -619,7 +807,8 @@ def test_build_demux_command(self):
             self.assertTrue(str(self.untrimmed_dir_fmt) in obs[10])
             self.assertEqual(str(self.seqs_dir_fmt.file.view(FastqGzFormat)),
                              obs[11])
-            self.assertTrue('1' in obs[13])
+            self.assertTrue('0' in obs[13])
```
Comment (colinvwood) [inline q2_cutadapt/tests/test_demux.py]:
```suggestion
            self.assertTrue('0' in obs[13])  # fwd cut
```

File: q2_cutadapt/tests/test_demux.py [hunk 5]
```diff
@@ -619,7 +807,8 @@ def test_build_demux_command(self):
...
+            self.assertTrue('1' in obs[15])
```
Comment (colinvwood) [inline q2_cutadapt/tests/test_demux.py]:
```suggestion
            self.assertTrue('1' in obs[15])  # cores
```

File: q2_cutadapt/tests/test_demux.py [hunk 6]
```diff
@@ -712,6 +901,8 @@ def test_build_demux_command(self):
         self.assertEqual(exp_f, obs[15])
         exp_r = str(self.seqs_dir_fmt.reverse_sequences.view(FastqGzFormat))
         self.assertEqual(exp_r, obs[16])
+        self.assertEqual('0', obs[18])
```
Comment (colinvwood) [inline q2_cutadapt/tests/test_demux.py]:
```suggestion
        self.assertEqual('0', obs[18])  # rev cut
```

File: q2_cutadapt/tests/test_demux.py [hunk 7]
```diff
@@ -712,6 +901,8 @@ def test_build_demux_command(self):
...
+        self.assertEqual('0', obs[20])
```
Comment (colinvwood) [inline q2_cutadapt/tests/test_demux.py]:
```suggestion
        self.assertEqual('0', obs[20])  # fwd cut
        self.assertEqual('1', obs[22])  # cores
```

**Comment by colinvwood** [comment]
Hi @vaamb,

I just added some comments in the command tests because they're getting pretty confusing. I believe I got the order correctly, but you can double check if you want. Also added the cores test for paired demux since you added it for single demux.

**Comment by vaamb** [comment]
Hi @colinvwood ,

Good idea to add those comments, it was indeed becoming confusing. I doubled check them and as far I can tell the orders are correct.

**Comment by colinvwood** [comment]
Looks good to me, thanks for all your work on this @vaamb!

**Comment by vaamb** [comment]
Thanks a lot for the reviewing @colinvwood !
